### ISSUE TITLE
# Lecture 7. Getting Neural Nets to Train

### ISSUE THREAD

**Issue by seonghunn** [seonghunn/CS182#7]
# Lecture 7. Getting Neural Nets to Train
## Part 1
<img width="293" alt="image" src="https://user-images.githubusercontent.com/69765278/165922758-5a5c83ed-cf0f-4285-aa70-763a1c8940e8.png">
위와 같이, NN을 잘 트레이닝 할 수 있는 여러 테크닉에 대해 다룬다.

<img width="651" alt="image" src="https://user-images.githubusercontent.com/69765278/165922913-d3ba0094-b7d3-4a90-b304-985f05be8f9c.png">
2D data에서, 위와 같이 한 쪽 dimension이 상대적으로 매우 클 경우(스케일의 차이가 있는 경우), 트레이닝 하기 어려워진다.

<img width="596" alt="image" src="https://user-images.githubusercontent.com/69765278/165923434-51059bf0-a17b-4194-9fa5-f9a430c15856.png">
위와 같이 backpropagation을 이용하여 dL/dW<sup>z</sup> 을 수행하면 이 결과가 dimension끼리 너무 차이가 나고, gradient가 왼쪽 (easy case)에서는 직접적으로 optimum을 가리키지만, hard case의 경우는 그렇지 못해 조금 복잡하게 optimum을 찾아가야 한다.

<img width="590" alt="image" src="https://user-images.githubusercontent.com/69765278/165924546-8b388fe4-3d76-4a8b-8a17-9cad79508d01.png">
이를 해결하기 위해 데이터의 평균을 0 표준편차를 1로 만드는 작업(standardization)을 하게 된다. 이 때 데이터가 소실되는 것은 아니기 때문에 좋은 방법이라고 할 수 있다.

<img width="565" alt="image" src="https://user-images.githubusercontent.com/69765278/165925617-13b39743-21bc-4ce1-87cc-7ef094f255b0.png">
위와 같이 activation 단계에서 dimension이 차이가 나는 문제가 발생한다면 input뿐 아니라 activation도 standardization을 수행할 수 있다.

<img width="694" alt="image" src="https://user-images.githubusercontent.com/69765278/165925517-a6d82dd3-5299-43f9-be12-00b6119e30f8.png">
이 방법의 단점은 평균과 표준편차가 W와 b값에 따라 달라지므로 트레이닝 과정에서 저 값들을 변경할 때마다 평균과 표준편차를 새로 구해야 한다는 것이다.

<img width="600" alt="image" src="https://user-images.githubusercontent.com/69765278/165926387-1873a74b-39ba-450c-acee-1b5c953b9834.png">
이를 해결하기 위해 위와 같이 batch normalization을 사용하게 되는데 이는 Batch에 대해서만 평균과 표준편차를 구하는 방법이다. 그리고 &gamma; 와 &beta; 값을 이용해 원래의 ai dimension으로 맞춘다.

<img width="474" alt="image" src="https://user-images.githubusercontent.com/69765278/165927141-03e5555b-d5b8-48d6-bd91-c98ab1cfe4c0.png">
이 normalization을 어디서 수행할 지에 대한 문제가 있는데, non linear layer 앞, 뒤에서 전부 동작한다.

batch normalization을 사용할 때 고려할 사항으로는
- 큰 learning rate를 사용할 수 있는 경우가 많음
- model이 더 빠르게 train됨
- Regularization이 덜 필요함
등이 있다.

이를 사용할 때, test 단계에서는 training set에서 구했던 평균과 표준편차를 그대로 이용하게 된다.

#Part 2
##Weight initialization
<img width="522" alt="image" src="https://user-images.githubusercontent.com/69765278/166128673-64fef393-f9fb-45d1-a964-13402facb051.png">
위와 같이 gradient가 여러 값의 곱으로 이루어지기 때문에, 이것이 너무 크거나 너무 작지 않도록 해 주어야 한다.
이렇게 되려면, 각 jacobian 값이 1에 근접해야 한다.

이렇게 만들기 위해서는 가장 간단히 가우시안 분포에서 구한 random weight를 사용할 수 있다.
하지만 이렇게 한다면 1보다 작은 값을 계속 곱하기 때문에 deep layer로 갈수록 activiation이 0에 매우 근접해지는 결과를 얻게 되고 이렇게 되면 gradient가 0에 가까워지기 때문에 plateau 등에 갇히는 결과를 만들어 낼수도 있다.

###Xavier initialization
<img width="602" alt="image" src="https://user-images.githubusercontent.com/69765278/166128921-c0223e87-6bdf-442f-986e-1c100f0eab4f.png">
<img width="345" alt="image" src="https://user-images.githubusercontent.com/69765278/166128928-e284e80c-bacd-4357-911e-176a950f86c2.png">
앞서 언급한 것과 같이 activation이 shirink 하거나 grow하는 것을 막기 위해 variation 공식을 사용한 E{z<sub>i</sub><sup>2</sup>} 값을 통해, w의 std distribution을 1/sqrt Do 로 정해주는 것이 xavier initialization 이라고 한다.

하지만 활성화 함수를 고려하지 못한 방법인데, ReLU를 예로 들 시, ReLU는 절반의 값이 0이기 때문에 계속 ReLU를 거치게 되면 값이 계속 0으로 수렴하게 된다.

<img width="685" alt="image" src="https://user-images.githubusercontent.com/69765278/166129052-e396644b-d902-4484-b798-34f7cf7b3fce.png">

따라서 위와 같이 표준편차를 조금 키워서 설정화는 방법으로 해결한다. 이는 very deep한 모델일 때 효과가 잘 나타나게 된다.

그리고, bias가 0이라면 ReLU에서 절반의 값은 0으로 죽어버리기 때문에, 작은 상수값으로 bias를 설정해 주어야 한다.

###Advanced initialization
<img width="564" alt="image" src="https://user-images.githubusercontent.com/69765278/166129155-51034747-c36a-4154-9862-d752082afc3e.png">

여러 jacobian의 곱으로 이루어진 gradient의 jacobian을 각각 SVD를 통해 orthonormal basis와 eigenvalues의 곱으로 분리해 낼 수 있다. 이 때 U와 V는 orthonormal basis로 scale을 보존한 채 linear transformation으로 동작하지만, Lamda 부분은 eigenvalue로 linear transformation의 결과가 rotation이 아니고 scaling이 되게 된다.

<img width="640" alt="image" src="https://user-images.githubusercontent.com/69765278/166129255-a81290a6-6f9c-41ac-8807-5a59f62460e9.png">
이 때, lamda 부분을 Identity matrix가 되게 하면 1에 가까운 scaling 값을 얻을 수 있다.

우리가 training을 수행할 때, gradient의 값이 매우 튀는 monster gradient가 나타날 수 있다. 이는 다음과 같은 원인 때문에 나타난다.

<img width="324" alt="image" src="https://user-images.githubusercontent.com/69765278/166129341-55c91b3a-d076-4b9d-8d53-24c4d052a04f.png">

<img width="445" alt="image" src="https://user-images.githubusercontent.com/69765278/166129351-3fa499a8-efaa-423f-9483-03530618f940.png">

이를 해결하기 위해 위와 같은 "clipping"이라는 방법을 사용하여 gradient 값을 정해주는 방법을 사용할 수 있다.
이 때 gradient 값을 보정하는 c 값은 몇 번의 epoch를 수행해 보고 바람직한 magnitude를 확인한 뒤 그에 맞게 정해주면 된다.

#Part 3
<img width="660" alt="image" src="https://user-images.githubusercontent.com/69765278/166130290-09fb1575-b3d3-4eeb-8ab4-cb0cc552a995.png">

위와 같이 모델이 실수를 했을 때, 이를 고치기 위한 방법 중 하나가 여러 가지 모델을 써서 평균값을 사용하는 방법인 앙상블을 사용하게 된다.

<img width="641" alt="image" src="https://user-images.githubusercontent.com/69765278/166130301-211e2b64-83ea-4969-855b-5d8d80e6c9e4.png">

D라는 dataset에서 M개의 다른 datasets를 구성할 때, overlapping하지만 independently하게 sample하게 되면 i.i.d하게 데이터셋을 구성할 수 있다. 즉 다음과 같은 방법으로 resampling 하는 것이다.

<img width="438" alt="image" src="https://user-images.githubusercontent.com/69765278/166130331-e0c4d7b9-14e5-4a7e-a808-80c0e15264ab.png">

<img width="355" alt="image" src="https://user-images.githubusercontent.com/69765278/166130347-c80191b2-ca4a-47aa-8789-fef363e23e51.png">

그렇게, 각각의 datasets를 서로 다른 모델에서 training 시키고 그 parameter 값들을 평균을 내거나 majority vote 방법을 이용하여 얻게 된다.
majority vote를 사용하지 않는 이유는 개가 정답인 모델에서 늑대가 50퍼, 개가 45퍼가 나왔다면 개의 45퍼라는 데이터는 아예 반영되지 않기 때문에 wrong answer에 보다 취약해진다는 단점 때문이다. 따라서 평균을 이용하는 것이 조금 더 바람직하다.

NN에는 랜덤성이 너무 많다. initialization, batch, stochastic gradient descent 등등.. 따라서 앙상블을 통해 랜덤성을 극복할 수 있고, 많은 모델을 학습시켜 더 오래걸린다고 생각할 수 있지만

<img width="566" alt="image" src="https://user-images.githubusercontent.com/69765278/166130408-a53c9fa2-ec68-43f2-9a69-a265ed5e1cf8.png">

위와 같이 task-agnostic한 부분(task에 불변인 부분)은 feature를 공유하고, task-specific인 부분만 각 모델에서 따로 처리하는 방식으로 속도를 높일 수 있다.

<img width="566" alt="image" src="https://user-images.githubusercontent.com/69765278/166130441-26c83759-9819-4eaa-80dc-37ce750b4e7a.png">

더더욱 빠른 방법은 위와 같은 snapshot ensemble라고 부르는데 parameter snapshot를 저장하는 방법으로, 이 snapshot를 앙상블 과정에서 계속 사용하는 것이다. 이를 합칠 때에는, 앞선 방법처럼 평균 확률, majority vote 방법 외에도 모든 parameter vector를 합치는 방법도 있다.

다음으로 대규모 앙상블이 대개 더 효과적이기 때문에, 이를 위해 랜덤하게 몇 개의 activation을 0으로 만드는 방법을 사용한다. 이를 dropout이라고 한다.

<img width="566" alt="image" src="https://user-images.githubusercontent.com/69765278/166130503-e57ea16a-102c-4229-99a1-c0597ee6c80b.png">

이것이 대규모 앙상블이라고 할 수 있는 이유는, 몇 개의 activation이 0으로 되는 것이 다른 모델을 정의한 것으로 생각할 수 있기 때문이다.

<img width="400" alt="image" src="https://user-images.githubusercontent.com/69765278/166130544-486c0a7c-750f-4502-a811-adaa8a469909.png">

따라서, training을 할 때에는 dropout을 위해 베르누이 분포를 activation에 곱해주고 test를 할 때는 dropout를 사용하지 않기 때문에 W가 2배 더 큰 값으로 되어 있는 것을 보정하기 위해 (dropout 과정의 결과에서) 1/2을 곱해줘 보정한다.

우리가 hyperparameter에서 배운 것들을 정리하면 다음과 같다.

<img width="665" alt="image" src="https://user-images.githubusercontent.com/69765278/166130573-be3005c0-6e29-4751-bffe-84fe0e889942.png">
